src/index.js

```diff
@@ -53,41 +53,75 @@ export const normalize = (input, schema) => {
   return { entities, result };
 };
 
-const unvisit = (input, schema, getDenormalizedEntity) => {
-  if (typeof schema === 'object' && (!schema.denormalize || typeof schema.denormalize !== 'function')) {
-    const method = Array.isArray(schema) ? ArrayUtils.denormalize : ObjectUtils.denormalize;
-    return method(schema, input, unvisit, getDenormalizedEntity);
+const unvisitEntity = (input, schema, unvisit, getEntity, cache) => {
+  const entity = getEntity(input, schema);
+  if (typeof entity !== 'object' || entity === null) {
+    return entity;
   }
 
-  if (input === undefined || input === null) {
-    return input;
+  const id = schema.getId(entity);
+
+  if (!cache[schema.key]) {
+    cache[schema.key] = {};
   }
 
-  return schema.denormalize(input, unvisit, getDenormalizedEntity);
-};
+  if (!cache[schema.key][id]) {
+    // Ensure we don't mutate it non-immutable objects
+    const entityCopy = ImmutableUtils.isImmutable(entity) ? entity : { ...entity };
 
-const getEntity = (entityOrId, schemaKey, entities, isImmutable) => {
-  if (typeof entityOrId === 'object') {
-    return entityOrId;
+    // Need to set this first so that if it is referenced further within the
+    // denormalization the reference will already exist.
+    cache[schema.key][id] = entityCopy;
+    cache[schema.key][id] = schema.denormalize(entityCopy, unvisit);
   }
 
-  return isImmutable ?
-    entities.getIn([ schemaKey, entityOrId.toString() ]) :
-    entities[schemaKey][entityOrId];
+  return cache[schema.key][id];
 };
 
-const getEntities = (entities, isImmutable) => (schema, entityOrId) => {
-  const schemaKey = schema.key;
+const getUnvisit = (entities) => {
+  const cache = {};
+  const getEntity = getEntities(entities);
+
+  const unvisit = (input, schema) => {
```

Comment:
Given that we need to recursively call `unvisit` within itself it needs to be named so we can't return an anonymous function. However, we can just return a named function which I think addresses this.